### PR TITLE
fix(aws service): more precise aws retry logic

### LINKED ISSUE #9005: `aws_sqs` sink does not retry throttled events
Reported by user in discord: https://discord.com/channels/742820443487993987/759154784309215252/882846307616702484

```
ERROR sink{component_kind="sink" component_name=metrics_to_sqs component_type=aws_sqs}:request{request_id=2555}: vector::sinks::util::retries: Non-retriable error; dropping the request. error=Request ID: Some("5ec64e84-89ac-58a3-8f60-172edc7095c5") Body: <?xml version="1.0"?><ErrorResponse xmlns="http://queue.amazonaws.com/doc/2012-11-05/"><Error><Type>Sender</Type><Code>RequestThrottled</Code><Message>Request is throttled</Message><Detail/></Error><RequestId>5ec64e84-89ac-58a3-8f60-172edc7095c5</RequestId></ErrorResponse
```

Vector should be considering this a retriable error.

We should check the other AWS sinks as well to ensure they retry throttling errors.

### PR DESCRIPTION
**Pull request by lukesteensen**
Closes #9005

I can't say that this is exhaustively tested, but I went through the API docs and some official client implementations to try to dig out some general patterns that we should be able to rely on here.

The most questionable part is the simple string matching instead of actually parsing the body, but the unknown format and structure made that seem tricky. It feels unlikely these error codes would be present in a 400-level response but not be _the_ error code.

I'm also not terribly attached to the mechanics of the matching. Given that the body size could be potentially large, it seemed worthwhile to match in a single pass, hence the `RegexSet`. But that may not be true with the additional complexity of the `OnceCell` and the fact that these will only happen occasionally and alongside a far more expensive HTTP request. I did at least make sure not to use `lazy_static` for @blt 😄

### CODE REVIEW

**Review by spencergilbert — APPROVED** [review]

File: src/rusoto/mod.rs [hunk 1]
```diff
@@ -339,14 +341,48 @@ impl From<Option<SignedRequestPayload>> for RusotoBody {
...
+            let re = RETRIABLE_CODES.get_or_init(|| {
+                RegexSet::new(&["RequestTimeout", "RequestExpired", "ThrottlingException"])
+                    .expect("invalid regex")
+            });
+            if response.status.is_client_error() && re.is_match(&response.body) {
+                return true;
+            }
+
+            response.status.is_server_error()
+                || response.status == http::StatusCode::TOO_MANY_REQUESTS
```
Comment (jszwedko) [inline src/rusoto/mod.rs:385]: Should we check this before doing the body matching? It seems like a less expensive check.

**Review by jszwedko — APPROVED** [review]
Nice! It might be nice to wrap some tests around this given how the logic here has expanded but it is is pretty straight-forward so not-blocking.

It looks like the new AWS SDK includes a retry policy we can use in the future: https://awslabs.github.io/aws-sdk-rust/aws_http/struct.AwsErrorRetryPolicy.html

**Comment by blt** [comment]
I fully approve of this usage of `OnceCell`.

File: src/rusoto/mod.rs (hunk 1 above)
Comment (lukesteensen) [reply src/rusoto/mod.rs:385]: Yep, good call.

File: src/rusoto/mod.rs [hunk 2]
```diff
@@ -339,14 +341,49 @@ impl From<Option<SignedRequestPayload>> for RusotoBody {
...
-        RusotoError::Unknown(res)
-            if res.status.is_server_error()
-                || res.status == http::StatusCode::TOO_MANY_REQUESTS =>
-        {
-            true
+        RusotoError::Unknown(response) => {
+            // This header is a direct indication that we should retry the request. Eventually it'd
+            // be nice to actually schedule the retry after the given delay, but for now we just
+            // check that it contains a positive value.
+            if response
```
Comment (bruceg) [inline src/rusoto/mod.rs:353]: This could be reworded as an `if` … `else if` … `else` block and avoid the returns.

**Review by bruceg — APPROVED** [review]
LGTM, just one comment about "code smell", not blocking

File: src/rusoto/mod.rs (hunk 2 above)
Comment (lukesteensen) [reply src/rusoto/mod.rs]: Good idea. I actually went one step further and just expressed the conditional directly.

**Pull request merged**